File: client/src/modules/assetmanager/schemas/captable/security-transaction.schemas.ts
/**
 * Security Transaction Schemas
 *
 * Zod validation schemas for SecurityTransaction model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/captable_models.py
 * - Schema: /server/apps/assetmanager/schemas/captable_schemas/security_transaction_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/captable_subrouters/security_transaction_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

/**
 * Transaction type options - matches backend TransactionType enum
 * Backend: class TransactionType(str, Enum)
 */
export const TransactionTypeEnum = z.enum([
  'issuance',
  'transfer',
  'conversion',
  'redemption',
  'exercise',
  'cancellation',
  'split',
  'merger',
]);

// ==========================================
// SecurityTransaction Schema (Full Representation)
// ==========================================

/**
 * SecurityTransaction schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class SecurityTransaction(BaseModel)
 */
export const SecurityTransactionSchema = z.object({
  id: z.number(),
  entity_id: z.number(),
  stakeholder_id: z.number(),
  funding_round_id: z.number(),
  security_id: z.number().nullable(),
  transaction_reference: z.string().max(50, 'Transaction reference must be less than 50 characters'),
  transaction_type: TransactionTypeEnum,
  units_debit: z.number(),
  units_credit: z.number(),
  amount_debit: z.number(),
  amount_credit: z.number(),
  transaction_date: z.string(), // ISO date string from backend
  notes: z.string().nullable(),
  related_transaction_id: z.number().nullable(),
  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new security transaction (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class SecurityTransactionCreate(BaseModel)
 */
export const CreateSecurityTransactionSchema = z.object({
  // Required fields
  entity_id: z.number(),
  stakeholder_id: z.number(),
  funding_round_id: z.number(),
  transaction_reference: z.string().min(1, 'Transaction reference is required').max(50, 'Transaction reference must be less than 50 characters'),
  transaction_type: TransactionTypeEnum,
  transaction_date: z.string().min(1, 'Transaction date is required'),

  // Optional fields
  security_id: z.number().nullable().optional(),
  units_debit: z.number().optional().default(0),
  units_credit: z.number().optional().default(0),
  amount_debit: z.number().optional().default(0),
  amount_credit: z.number().optional().default(0),
  notes: z.string().nullable().optional(),
  related_transaction_id: z.number().nullable().optional(),
});

/**
 * Schema for updating a security transaction (PUT)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class SecurityTransactionUpdate(BaseModel)
 */
export const UpdateSecurityTransactionSchema = z.object({
  entity_id: z.number().optional(),
  stakeholder_id: z.number().optional(),
  funding_round_id: z.number().optional(),
  security_id: z.number().nullable().optional(),
  transaction_reference: z.string().max(50, 'Transaction reference must be less than 50 characters').optional(),
  transaction_type: TransactionTypeEnum.optional(),
  units_debit: z.number().optional(),
  units_credit: z.number().optional(),
  amount_debit: z.number().optional(),
  amount_credit: z.number().optional(),
  transaction_date: z.string().optional(),
  notes: z.string().nullable().optional(),
  related_transaction_id: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type TransactionType = z.infer<typeof TransactionTypeEnum>;
export type SecurityTransaction = z.infer<typeof SecurityTransactionSchema>;
export type CreateSecurityTransaction = z.infer<typeof CreateSecurityTransactionSchema>;
export type UpdateSecurityTransaction = z.infer<typeof UpdateSecurityTransactionSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single security transaction
 * Backend equivalent: class SecurityTransactionResponse(BaseModel)
 */
export type SecurityTransactionResponse = {
  success: boolean;
  data?: SecurityTransaction;
  error?: string;
};

/**
 * Response containing multiple security transactions
 * Backend equivalent: class SecurityTransactionsResponse(BaseModel)
 */
export type SecurityTransactionsResponse = {
  success: boolean;
  data?: SecurityTransaction[];
  error?: string;
};

// ==========================================
// Helper Functions (adapted from v7capital)
// ==========================================

/**
 * Human-readable labels for transaction types
 */
export const TRANSACTION_TYPE_LABELS: Record<TransactionType, string> = {
  issuance: 'Issuance',
  transfer: 'Transfer',
  conversion: 'Conversion',
  redemption: 'Redemption',
  exercise: 'Exercise',
  cancellation: 'Cancellation',
  split: 'Split',
  merger: 'Merger',
};

/**
 * Get human-readable label for a transaction type
 * @param type Transaction type enum value
 * @returns Display label string
 */
export const getTransactionTypeLabel = (type: TransactionType): string => {
  return TRANSACTION_TYPE_LABELS[type] || type;
};

/**
 * Calculate net amount for a transaction (credit minus debit)
 * Positive = inflow, Negative = outflow
 * @param transaction Security transaction
 * @returns Net amount
 */
export const calculateNetAmount = (transaction: SecurityTransaction): number => {
  return transaction.amount_credit - transaction.amount_debit;
};

/**
 * Calculate net units for a transaction (credit minus debit)
 * Positive = units received, Negative = units sent
 * @param transaction Security transaction
 * @returns Net units
 */
export const calculateNetUnits = (transaction: SecurityTransaction): number => {
  return transaction.units_credit - transaction.units_debit;
};

File: server/apps/assetmanager/schemas/captable_schemas/security_transaction_schemas.py
"""
SecurityTransaction Schemas

Pydantic schemas for the SecurityTransaction model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type
from enum import Enum

class TransactionType(str, Enum):
    """Security transaction type options"""
    ISSUANCE = "issuance"
    TRANSFER = "transfer"
    CONVERSION = "conversion"
    REDEMPTION = "redemption"
    EXERCISE = "exercise"
    CANCELLATION = "cancellation"
    SPLIT = "split"
    MERGER = "merger"

# ==========================================
# SecurityTransaction Schema (Full Representation)
# ==========================================

class SecurityTransaction(BaseModel):
    """SecurityTransaction schema - full representation"""
    id: int
    entity_id: int = Field(description="Associated entity ID")
    stakeholder_id: int = Field(description="Associated stakeholder ID")
    funding_round_id: int = Field(description="Associated funding round ID")
    security_id: int | None = Field(None, description="Associated security ID (NULL for fund-level transactions)")
    transaction_reference: str = Field(max_length=50, description="Transaction reference")
    transaction_type: TransactionType = Field(description="Transaction type")
    units_debit: float = Field(default=0, description="Units debit amount")
    units_credit: float = Field(default=0, description="Units credit amount")
    amount_debit: float = Field(default=0, description="Cash debit amount")
    amount_credit: float = Field(default=0, description="Cash credit amount")
    transaction_date: date_type = Field(description="Transaction date")
    notes: str | None = Field(None, description="Transaction notes")
    related_transaction_id: int | None = Field(None, description="Related transaction ID")
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class SecurityTransactionCreate(BaseModel):
    """Schema for creating a new security transaction"""
    entity_id: int = Field(description="Associated entity ID")
    stakeholder_id: int = Field(description="Associated stakeholder ID")
    funding_round_id: int = Field(description="Associated funding round ID")
    security_id: int | None = Field(None, description="Associated security ID (NULL for fund-level transactions)")
    transaction_reference: str = Field(max_length=50, description="Transaction reference")
    transaction_type: TransactionType = Field(description="Transaction type")
    units_debit: float = Field(default=0, description="Units debit amount")
    units_credit: float = Field(default=0, description="Units credit amount")
    amount_debit: float = Field(default=0, description="Cash debit amount")
    amount_credit: float = Field(default=0, description="Cash credit amount")
    transaction_date: date_type = Field(description="Transaction date")
    notes: str | None = Field(None, description="Transaction notes")
    related_transaction_id: int | None = Field(None, description="Related transaction ID")

class SecurityTransactionUpdate(BaseModel):
    """Schema for updating a security transaction"""
    entity_id: int | None = None
    stakeholder_id: int | None = None
    funding_round_id: int | None = None
    security_id: int | None = None
    transaction_reference: str | None = Field(None, max_length=50)
    transaction_type: TransactionType | None = None
    units_debit: float | None = None
    units_credit: float | None = None
    amount_debit: float | None = None
    amount_credit: float | None = None
    transaction_date: date_type | None = None
    notes: str | None = None
    related_transaction_id: int | None = None

# ==========================================
# Response Types
# ==========================================

class SecurityTransactionResponse(BaseModel):
    """Response containing a single security transaction"""
    success: bool
    data: SecurityTransaction | None = None
    error: str | None = None

class SecurityTransactionsResponse(BaseModel):
    """Response containing multiple security transactions"""
    success: bool
    data: list[SecurityTransaction] | None = None
    error: str | None = None